<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Optimizer configuration · COBREXA.jl</title><meta name="title" content="Optimizer configuration · COBREXA.jl"/><meta property="og:title" content="Optimizer configuration · COBREXA.jl"/><meta property="twitter:title" content="Optimizer configuration · COBREXA.jl"/><meta name="description" content="Documentation for COBREXA.jl."/><meta property="og:description" content="Documentation for COBREXA.jl."/><meta property="twitter:description" content="Documentation for COBREXA.jl."/><meta property="og:url" content="https://cobrexa.github.io/COBREXA.jl/stable/optimizers/"/><meta property="twitter:url" content="https://cobrexa.github.io/COBREXA.jl/stable/optimizers/"/><link rel="canonical" href="https://cobrexa.github.io/COBREXA.jl/stable/optimizers/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img class="docs-light-only" src="../assets/logo.svg" alt="COBREXA.jl logo"/><img class="docs-dark-only" src="../assets/logo-dark.svg" alt="COBREXA.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../">COBREXA.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><span class="tocitem">Home</span><ul><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../structure/">Core concepts and structure</a></li><li class="is-active"><a class="tocitem" href>Optimizer configuration</a><ul class="internal"><li><a class="tocitem" href="#Optimizer-choice-considerations"><span>Optimizer choice considerations</span></a></li><li><a class="tocitem" href="#Optimizer-parameters"><span>Optimizer parameters</span></a></li></ul></li></ul></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../examples/">Contents</a></li><li><a class="tocitem" href="../examples/01-loading-and-saving/">Loading and saving models</a></li><li><a class="tocitem" href="../examples/02a-flux-balance-analysis/">Flux balance analysis (FBA)</a></li><li><a class="tocitem" href="../examples/02b-optimizer-parameters/">Changing optimizer parameters</a></li><li><a class="tocitem" href="../examples/02c-model-modifications/">Making adjustments to the model</a></li><li><a class="tocitem" href="../examples/02d-constraint-modifications/">Making adjustments to the constraint system</a></li><li><a class="tocitem" href="../examples/03a-flux-variability-analysis/">Flux variability analysis (FVA)</a></li><li><a class="tocitem" href="../examples/03b-parsimonious-flux-balance/">Parsimonious flux balance analysis</a></li><li><a class="tocitem" href="../examples/03c-envelopes/">Production envelopes</a></li><li><a class="tocitem" href="../examples/03d-unidirectional/">Split unidirectional reactions in models</a></li><li><a class="tocitem" href="../examples/03e-min-exchange/">Parsimonious growth medium</a></li><li><a class="tocitem" href="../examples/04-community-models/">Community FBA models</a></li><li><a class="tocitem" href="../examples/04a-composition-models/">Steady community composition models</a></li><li><a class="tocitem" href="../examples/05a-minimization-of-metabolic-adjustment/">Minimization of metabolic adjustment analysis</a></li><li><a class="tocitem" href="../examples/05b-enzyme-constrained-models/">Enzyme constrained models</a></li><li><a class="tocitem" href="../examples/05c-mmdf/">Thermodynamic models</a></li><li><a class="tocitem" href="../examples/05d-loopless-models/">Loopless flux balance analysis (ll-FBA)</a></li><li><a class="tocitem" href="../examples/05e-knockouts/">Gene knockouts</a></li><li><a class="tocitem" href="../examples/05f-cyclefree/">CycleFreeFlux</a></li><li><a class="tocitem" href="../examples/05g-gapfilling/">Gap filling</a></li><li><a class="tocitem" href="../examples/05h-medium/">Growth medium optimization</a></li><li><a class="tocitem" href="../examples/06a-sampling/">Flux sampling</a></li><li><a class="tocitem" href="../examples/06b-screening/">Screening through many model variants</a></li><li><a class="tocitem" href="../examples/07a-srba/">Simplified resource balance analysis</a></li><li><a class="tocitem" href="../examples/07b-community-ecfba/">Enzyme-constrained communities</a></li></ul></li><li><span class="tocitem">Parallel, distributed and HPC processing</span><ul><li><a class="tocitem" href="../distributed/">Contents</a></li><li><a class="tocitem" href="../distributed/1_functions/">Parallel processing overview</a></li><li><a class="tocitem" href="../distributed/2_parallel/">Local parallel processing</a></li><li><a class="tocitem" href="../distributed/3_slurm/">Working in a HPC environment</a></li></ul></li><li><span class="tocitem">API reference</span><ul><li><a class="tocitem" href="../reference/">Contents</a></li><li><a class="tocitem" href="../reference/analysis/">Specialized analysis functions</a></li><li><a class="tocitem" href="../reference/builders/">Constraint system builders</a></li><li><a class="tocitem" href="../reference/core/">Core functionality</a></li><li><a class="tocitem" href="../reference/frontend/">Front-end user interface</a></li><li><a class="tocitem" href="../reference/misc/">Miscellaneous functions</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Home</a></li><li class="is-active"><a href>Optimizer configuration</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Optimizer configuration</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/COBREXA/COBREXA.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/COBREXA/COBREXA.jl/blob/master/docs/src/optimizers.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Optimizer-compatibility-and-settings"><a class="docs-heading-anchor" href="#Optimizer-compatibility-and-settings">Optimizer compatibility and settings</a><a id="Optimizer-compatibility-and-settings-1"></a><a class="docs-heading-anchor-permalink" href="#Optimizer-compatibility-and-settings" title="Permalink"></a></h1><p>COBREXA uses <a href="https://github.com/jump-dev/JuMP.jl">JuMP.jl</a> as an abstraction layer above the Julia optimization ecosystem and <a href="https://github.com/jump-dev/MathOptInterface.jl">MathOptInterface.jl</a>. Problems represented in COBREXA are typically not implemented in a optimizer-specific manner, and it is very easy to switch between different optimizers and various configurations to try which ones are able to represent the model best. Generally, you can install any optimizer from <a href="https://jump.dev/JuMP.jl/stable/installation/#Supported-solvers">JuMP&#39;s list</a> and work with it.</p><p>For example, to install SCIP, we request it from Julia packages:</p><pre><code class="nohighlight hljs">] add SCIP</code></pre><p>...after which we can use <code>SCIP.Optimizer</code> object as a parameter to anything that requires the optimizer:</p><pre><code class="language-julia hljs">using SCIP
flux_balance_analysis(some_model, optimizer = SCIP.Optimizer)</code></pre><h2 id="Optimizer-choice-considerations"><a class="docs-heading-anchor" href="#Optimizer-choice-considerations">Optimizer choice considerations</a><a id="Optimizer-choice-considerations-1"></a><a class="docs-heading-anchor-permalink" href="#Optimizer-choice-considerations" title="Permalink"></a></h2><p>The desired optimizer is typically a subject to various constraints. Most notably, the optimizer must be capable of processing the optimization problem at hand. The choices reduce as follows:</p><ul><li>For analyses that require quadratic objectives (typically the <a href="../examples/03b-parsimonious-flux-balance/">L2-parsimonious analyses</a>), an optimizer with QP support is required.</li><li><a href="../reference/frontend/#COBREXA.gap_filling_analysis-Tuple">Gap-filling</a>, <a href="../reference/frontend/#COBREXA.medium_optimization_analysis-Tuple">medium optimization</a> and <a href="../reference/frontend/#COBREXA.loopless_flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}">loopless FBA</a> typically require MILP support in the optimizer.</li><li>In case you require numerically stable solutions, an IPM optimizer may be viable.</li></ul><p>Notably, some of the &quot;larger&quot; optimizers may support all of these options (these include SCIP, HiGHS, and the commercial Gurobi and CPLEX) at the slight cost of complexity and potential slowness in general cases.</p><h2 id="Optimizer-parameters"><a class="docs-heading-anchor" href="#Optimizer-parameters">Optimizer parameters</a><a id="Optimizer-parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Optimizer-parameters" title="Permalink"></a></h2><p>Most optimizer may be tuned by attributes, set via <code>settings</code> argument in most optimizing functions, and via optimizer attributes functionality of JuMP. In COBREXA, the attributes can be modified via <a href="../reference/misc/#COBREXA.set_optimizer_attribute-Tuple{Any, Any}"><code>set_optimizer_attribute</code></a>.</p><p>Unfortunately, the settings of individual optimizers differ quite significantly, so there is no good universal default and one has to consult the documentation of the given optimizer to find the actual names and expected values of attributes. On the other hand, following some simple guidelines may push most optimizers into behaving nearly optimally. We list the common considerations here:</p><ul><li>Tolerance settings do not have to be too strict in most cases; tolerance of <code>1e-5</code> is typically good for working with genome-scale models.<ul><li>In some cases, the tolerance has to be made stricter; e.g., gap-filling may &quot;enable&quot; a little amount of flux through an universal reaction that slips through the tolerance even without setting the corresponding binary variable to <code>1</code>, which in turn yields invalid solution. <strong>Tolerance settings in potentially numerically unstable MILP programs should always be validated</strong>.</li><li>In large models (e.g., the <a href="../examples/07b-community-ecfba/">enzyme-constrained communities</a>), individual in-tolerance errors may accumulate and harshly impact the precision of the solution. Apart from tightening the tolerance, some models may also require rescaling in order to be more numerically stable.</li></ul></li><li>Some analysis types (notably <a href="../reference/frontend/#COBREXA.flux_variability_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}">FVA</a>, <a href="../reference/frontend/#COBREXA.flux_sample-Tuple{AbstractFBCModels.AbstractFBCModel}">warm-up computation for samplers</a>, but also <a href="../reference/frontend/#COBREXA.gene_knockouts">gene knockouts</a> and <a href="../reference/frontend/#COBREXA.objective_production_envelope-Tuple{AbstractFBCModels.AbstractFBCModel, Vector{String}}">production envelopes</a>) benefit from the ability of the solver to quickly restart computation with only small modifications from the previous known solution (usually this is called a &quot;hot start&quot;). Especially for FVA, this brings great performance benefits. For such cases, one often has to force the solver to use a seemingly sub-optimal solving strategy that is sufficiently simple to allow the fast restarts.<ul><li>As a general rule of thumb, any kind of dual-solving method should be disabled for FVA and sampler warm-up, because the change of the optimization objective between algorithm iterations implies recomputation of the whole dual program. In particular <strong>for FVA, we obtained best results with the &quot;primal simplex&quot; solving methods</strong>.</li><li>For computations such as gene knockout analysis, the primal simplex is normally changed in each algorithm iteration, forcing a recomputation again. We obtained interesting results by using IPM solvers in such cases – the IPM solution often stays valid between different knockouts, speeding up the whole computation substantially.</li></ul></li><li>When solving linear programs in parallel (e.g., on a cluster or HPC), the additional parallelism introduced in the solver itself may cause scheduling issues, resource starvation and eventually general slowness. In such cases, reducing the number of solver threads to a single one may actually help the performance.</li></ul><p>For example, a good FVA-capable settings for HiGHS (gathered from <a href="https://ergo-code.github.io/HiGHS/dev/options/definitions/">its documentation</a>) may look like this:</p><pre><code class="language-julia hljs">import HiGHS
optimizer = HiGHS.Optimizer
settings = [
    set_optimizer_attribute(&quot;solver&quot;, &quot;simplex&quot;),
    set_optimizer_attribute(&quot;simplex_strategy&quot;, 4), # stands for &quot;primal simplex&quot;
    set_optimizer_attribute(&quot;parallel&quot;, &quot;off&quot;),
]

# ...

flux_variability_analysis(my_model; optimizer, settings)</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../structure/">« Core concepts and structure</a><a class="docs-footer-nextpage" href="../examples/">Contents »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.14.1 on <span class="colophon-date" title="Friday 11 July 2025 13:36">Friday 11 July 2025</span>. Using Julia version 1.11.6.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
