<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Loading and saving models ¬∑ COBREXA.jl</title><meta name="title" content="Loading and saving models ¬∑ COBREXA.jl"/><meta property="og:title" content="Loading and saving models ¬∑ COBREXA.jl"/><meta property="twitter:title" content="Loading and saving models ¬∑ COBREXA.jl"/><meta name="description" content="Documentation for COBREXA.jl."/><meta property="og:description" content="Documentation for COBREXA.jl."/><meta property="twitter:description" content="Documentation for COBREXA.jl."/><meta property="og:url" content="https://cobrexa.github.io/COBREXA.jl/stable/examples/01-loading-and-saving/"/><meta property="twitter:url" content="https://cobrexa.github.io/COBREXA.jl/stable/examples/01-loading-and-saving/"/><link rel="canonical" href="https://cobrexa.github.io/COBREXA.jl/stable/examples/01-loading-and-saving/"/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img class="docs-light-only" src="../../assets/logo.svg" alt="COBREXA.jl logo"/><img class="docs-dark-only" src="../../assets/logo-dark.svg" alt="COBREXA.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">COBREXA.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><span class="tocitem">Home</span><ul><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../../structure/">Core concepts and structure</a></li></ul></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../">Contents</a></li><li class="is-active"><a class="tocitem" href>Loading and saving models</a><ul class="internal"><li><a class="tocitem" href="#Getting-the-models-reliably-from-the-repositories"><span>Getting the models reliably from the repositories</span></a></li><li><a class="tocitem" href="#Loading-models"><span>Loading models</span></a></li><li><a class="tocitem" href="#Saving-models"><span>Saving models</span></a></li></ul></li><li><a class="tocitem" href="../02a-flux-balance-analysis/">Flux balance analysis (FBA)</a></li><li><a class="tocitem" href="../02b-optimizer-parameters/">Changing optimizer parameters</a></li><li><a class="tocitem" href="../02c-model-modifications/">Making adjustments to the model</a></li><li><a class="tocitem" href="../02d-constraint-modifications/">Making adjustments to the constraint system</a></li><li><a class="tocitem" href="../03a-flux-variability-analysis/">Flux variability analysis (FVA)</a></li><li><a class="tocitem" href="../03b-parsimonious-flux-balance/">Parsimonious flux balance analysis</a></li><li><a class="tocitem" href="../03c-envelopes/">Production envelopes</a></li><li><a class="tocitem" href="../03d-unidirectional/">Split unidirectional reactions in models</a></li><li><a class="tocitem" href="../04-community-models/">Community FBA models</a></li><li><a class="tocitem" href="../04a-composition-models/">Steady community composition models</a></li><li><a class="tocitem" href="../05a-minimization-of-metabolic-adjustment/">Minimization of metabolic adjustment analysis</a></li><li><a class="tocitem" href="../05b-enzyme-constrained-models/">Enzyme constrained models</a></li><li><a class="tocitem" href="../05c-mmdf/">Thermodynamic models</a></li><li><a class="tocitem" href="../05d-loopless-models/">Loopless flux balance analysis (ll-FBA)</a></li><li><a class="tocitem" href="../05e-knockouts/">Gene knockouts</a></li><li><a class="tocitem" href="../05f-cyclefree/">CycleFreeFlux</a></li><li><a class="tocitem" href="../05g-gapfilling/">Gap filling</a></li><li><a class="tocitem" href="../06a-sampling/">Flux sampling</a></li><li><a class="tocitem" href="../06b-screening/">Screening through many model variants</a></li><li><a class="tocitem" href="../07a-srba/">Simplified resource balance analysis</a></li><li><a class="tocitem" href="../07b-community-ecfba/">Enzyme-constrained communities</a></li></ul></li><li><span class="tocitem">Parallel, distributed and HPC processing</span><ul><li><a class="tocitem" href="../../distributed/">Contents</a></li><li><a class="tocitem" href="../../distributed/1_functions/">Parallel processing overview</a></li><li><a class="tocitem" href="../../distributed/2_parallel/">Local parallel processing</a></li><li><a class="tocitem" href="../../distributed/3_slurm/">Working in a HPC environment</a></li></ul></li><li><span class="tocitem">API reference</span><ul><li><a class="tocitem" href="../../reference/">Contents</a></li><li><a class="tocitem" href="../../reference/analysis/">Specialized analysis functions</a></li><li><a class="tocitem" href="../../reference/builders/">Constraint system builders</a></li><li><a class="tocitem" href="../../reference/core/">Core functionality</a></li><li><a class="tocitem" href="../../reference/frontend/">Front-end user interface</a></li><li><a class="tocitem" href="../../reference/misc/">Miscellaneous functions</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Examples</a></li><li class="is-active"><a href>Loading and saving models</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Loading and saving models</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/COBREXA/COBREXA.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands">ÔÇõ</span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/COBREXA/COBREXA.jl/blob/master/docs/src/examples/01-loading-and-saving.jl" title="Edit source on GitHub"><span class="docs-icon fa-solid">ÔÅÑ</span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Loading-and-saving-models"><a class="docs-heading-anchor" href="#Loading-and-saving-models">Loading and saving models</a><a id="Loading-and-saving-models-1"></a><a class="docs-heading-anchor-permalink" href="#Loading-and-saving-models" title="Permalink"></a></h1><pre><code class="language-julia hljs">using COBREXA</code></pre><h2 id="Getting-the-models-reliably-from-the-repositories"><a class="docs-heading-anchor" href="#Getting-the-models-reliably-from-the-repositories">Getting the models reliably from the repositories</a><a id="Getting-the-models-reliably-from-the-repositories-1"></a><a class="docs-heading-anchor-permalink" href="#Getting-the-models-reliably-from-the-repositories" title="Permalink"></a></h2><p>For convenience, COBREXA provides a specific function <a href="../../reference/core/#COBREXA.download_model-Tuple"><code>download_model</code></a> to download models from repositories that also automatically uses the cached downloaded version of the model if it&#39;s already downloaded, and verifies the checksum to improve reproducibility. It will print out a warning in case the model checksum does not match the expectation:</p><pre><code class="language-julia hljs">download_model(
    &quot;http://bigg.ucsd.edu/static/models/e_coli_core.json&quot;,
    &quot;e_coli_core.json&quot;,
    &quot;7bedec10576cfe935b19218dc881f3fb14f890a1871448fc19a9b4ee15b448d8&quot;,
)

download_model(
    &quot;http://bigg.ucsd.edu/static/models/e_coli_core.xml&quot;,
    &quot;e_coli_core.xml&quot;,
    &quot;b4db506aeed0e434c1f5f1fdd35feda0dfe5d82badcfda0e9d1342335ab31116&quot;,
)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">&quot;e_coli_core.xml&quot;</code></pre><div class="admonition is-success"><header class="admonition-header">How do I get the model hash?</header><div class="admonition-body"><p>We do not need to fill in the hash values immediately ‚Äì instead, it&#39;s possible to simply run the function once, and then copy the reported hash value from the warning message into the script.</p></div></div><h2 id="Loading-models"><a class="docs-heading-anchor" href="#Loading-models">Loading models</a><a id="Loading-models-1"></a><a class="docs-heading-anchor-permalink" href="#Loading-models" title="Permalink"></a></h2><p>To load genome-scale metabolic models, COBREXA uses the <a href="https://github.com/COBREXA/AbstractFBCModels.jl">AbstractFBCModels</a> framework to import various kinds of models including SBML, JSON and the legacy Matlab-formatted &quot;COBRA toolbox&quot; models.</p><p>All models can be loaded automatically using <a href="../../reference/core/#COBREXA.load_model-Tuple{String}"><code>load_model</code></a>; but one must import the model-type specific packages to load the functionality. (This step is required to keep the &quot;base&quot; COBREXA as efficient and fast-loading as possible.)</p><pre><code class="language-julia hljs">import JSONFBCModels, SBMLFBCModels

model1 = load_model(&quot;e_coli_core.json&quot;)

model2 = load_model(&quot;e_coli_core.xml&quot;)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">SBMLFBCModels.SBMLFBCModel(#= 95 reactions, 72 metabolites =#)</code></pre><p>We can now explore the contents of the models using the AbstractFBCModels&#39; interface:</p><pre><code class="language-julia hljs">import AbstractFBCModels as A

A.reactions(model1)

A.reactions(model2)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">95-element Vector{String}:
 &quot;R_ACALD&quot;
 &quot;R_ACALDt&quot;
 &quot;R_ACKr&quot;
 &quot;R_ACONTa&quot;
 &quot;R_ACONTb&quot;
 &quot;R_ACt2r&quot;
 &quot;R_ADK1&quot;
 &quot;R_AKGDH&quot;
 &quot;R_AKGt2r&quot;
 &quot;R_ALCD2x&quot;
 ‚ãÆ
 &quot;R_SUCCt2_2&quot;
 &quot;R_SUCCt3&quot;
 &quot;R_SUCDi&quot;
 &quot;R_SUCOAS&quot;
 &quot;R_TALA&quot;
 &quot;R_THD2&quot;
 &quot;R_TKT1&quot;
 &quot;R_TKT2&quot;
 &quot;R_TPI&quot;</code></pre><p>Additional extractable information can be found in <a href="https://cobrexa.github.io/AbstractFBCModels.jl/stable/reference/#Model-content-accessors">the documentation of the abstract models package</a>.</p><h3 id="Converting-model-types"><a class="docs-heading-anchor" href="#Converting-model-types">Converting model types</a><a id="Converting-model-types-1"></a><a class="docs-heading-anchor-permalink" href="#Converting-model-types" title="Permalink"></a></h3><p>Normally, <a href="../../reference/core/#COBREXA.load_model-Tuple{String}"><code>load_model</code></a> is forced to guess the model type from the filename suffix. We can specify the model type ourselves (this also allows the users to work with non-standard file suffixes, and saves some overhead and uncertainty in the guessing process):</p><pre><code class="language-julia hljs">import JSONFBCModels: JSONFBCModel

model = load_model(JSONFBCModel, &quot;e_coli_core.json&quot;)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">JSONFBCModels.JSONFBCModel(#= 95 reactions, 72 metabolites =#)</code></pre><p>Sometimes it is useful to convert the model data to another type, such as the SBML to a JSON model structure:</p><pre><code class="language-julia hljs">model_converted_to_json = load_model(&quot;e_coli_core.xml&quot;, JSONFBCModel)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">JSONFBCModels.JSONFBCModel(#= 95 reactions, 72 metabolites =#)</code></pre><p>Or to the &quot;Canonical Julia model&quot; from AbstractFBCModels:</p><pre><code class="language-julia hljs">model_in_julia_structures =
    load_model(JSONFBCModel, &quot;e_coli_core.json&quot;, A.CanonicalModel.Model)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">AbstractFBCModels.CanonicalModel.Model(
  reactions = Dict{String, AbstractFBCModels.CanonicalModel.Reaction}(&quot;ACALD&quot; =‚Ä¶
  metabolites = Dict{String, AbstractFBCModels.CanonicalModel.Metabolite}(&quot;glu_‚Ä¶
  genes = Dict{String, AbstractFBCModels.CanonicalModel.Gene}(&quot;b4301&quot; =&gt; Abstra‚Ä¶
  couplings = Dict{String, AbstractFBCModels.CanonicalModel.Coupling}(),
)
</code></pre><div class="admonition is-success"><header class="admonition-header">Compatibility with COBREXA v1.x</header><div class="admonition-body"><p><code>CanonicalModel</code> is a newer, cleaned-up version of the <code>StandardModel</code> type used in COBREXA version 1. For all code that relied on <code>StandardModel</code>, the canonical one should work just as well.</p></div></div><p>The above command specifies all model types explicitly, leaving least room for guessing-based errors.</p><p>If required, it is also possible to convert all model types to each other simply by using Julia&#39;s <code>convert</code>:</p><pre><code class="language-julia hljs">model_in_json_structure = convert(JSONFBCModel, model_in_julia_structures)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">JSONFBCModels.JSONFBCModel(#= 95 reactions, 72 metabolites =#)</code></pre><h2 id="Saving-models"><a class="docs-heading-anchor" href="#Saving-models">Saving models</a><a id="Saving-models-1"></a><a class="docs-heading-anchor-permalink" href="#Saving-models" title="Permalink"></a></h2><p>The models can be saved to file storage by using <a href="../../reference/core/#COBREXA.save_model-Union{Tuple{O}, Tuple{T}, Tuple{T, String, Type{O}}} where {T&lt;:AbstractFBCModels.AbstractFBCModel, O&lt;:AbstractFBCModels.AbstractFBCModel}"><code>save_model</code></a>:</p><pre><code class="language-julia hljs">save_model(model_converted_to_json, &quot;e_coli_core_from_sbml.json&quot;)</code></pre><p>Expectably, the file will contain the JSON with the model description:</p><pre><code class="language-julia hljs">println(open(&quot;e_coli_core_from_sbml.json&quot;) do f
    read(f, 100)
end |&gt; String, &quot;...&quot;)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">{&quot;metabolites&quot;:[{&quot;compartment&quot;:&quot;c&quot;,&quot;name&quot;:&quot;3-Phospho-D-glyceroyl phosphate&quot;,&quot;formula&quot;:&quot;C3H4O10P2&quot;,&quot;i...</code></pre><p>Note that without the conversion, it may happen that you save the model in an unexpected format!</p><pre><code class="language-julia hljs">save_model(model_in_julia_structures, &quot;e_coli_saved_wrongly.json&quot;)
println(open(&quot;e_coli_saved_wrongly.json&quot;) do f
    read(f, 100)
end |&gt; String, &quot;...&quot;)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">7JL   5Model|ì)å&amp;i!Ç¯@â˙=OZAbstractFBCModelsCanonicalModelD5DictNûD    !Reaction...</code></pre><p>The above code has saved the <code>CanonicalModel</code> in the way specified by the <code>CanonicalModel</code> structure ‚Äì which is, in this case, a binary dump of the Julia objects, instead of the expected JSON. To prevent this, you can either specify the output type yourself:</p><pre><code class="language-julia hljs">save_model(model_in_julia_structures, &quot;e_coli_saved_right.json&quot;, JSONFBCModel)</code></pre><p>...or use <a href="../../reference/core/#COBREXA.save_converted_model-Union{Tuple{T}, Tuple{T, String}} where T&lt;:AbstractFBCModels.AbstractFBCModel"><code>save_converted_model</code></a> to guess the model type automatically from the extension:</p><pre><code class="language-julia hljs">save_converted_model(model_in_julia_structures, &quot;e_coli_saved_automatically_right.json&quot;)
println(open(&quot;e_coli_saved_automatically_right.json&quot;) do f
    read(f, 100)
end |&gt; String, &quot;...&quot;)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">{&quot;metabolites&quot;:[{&quot;compartment&quot;:&quot;c&quot;,&quot;name&quot;:&quot;3-Phospho-D-glyceroyl phosphate&quot;,&quot;formula&quot;:&quot;C3H4O10P2&quot;,&quot;i...</code></pre><p>As with <a href="../../reference/core/#COBREXA.load_model-Tuple{String}"><code>load_model</code></a>, there is some overhead and uncertainty associated with <a href="../../reference/core/#COBREXA.save_converted_model-Union{Tuple{T}, Tuple{T, String}} where T&lt;:AbstractFBCModels.AbstractFBCModel"><code>save_converted_model</code></a> guessing the model type from extension. For that reason, it is adviseable to rely on the guessing functionality only in interactive use in REPL, and avoid it in automated scriptage altogether.</p><hr/><p><em>This page was generated using <a href="https://github.com/fredrikekre/Literate.jl">Literate.jl</a>.</em></p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">¬´ Contents</a><a class="docs-footer-nextpage" href="../02a-flux-balance-analysis/">Flux balance analysis (FBA) ¬ª</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.1 on <span class="colophon-date" title="Sunday 23 February 2025 12:37">Sunday 23 February 2025</span>. Using Julia version 1.11.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
