<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Front-end user interface · COBREXA.jl</title><meta name="title" content="Front-end user interface · COBREXA.jl"/><meta property="og:title" content="Front-end user interface · COBREXA.jl"/><meta property="twitter:title" content="Front-end user interface · COBREXA.jl"/><meta name="description" content="Documentation for COBREXA.jl."/><meta property="og:description" content="Documentation for COBREXA.jl."/><meta property="twitter:description" content="Documentation for COBREXA.jl."/><meta property="og:url" content="https://cobrexa.github.io/COBREXA.jl/stable/reference/frontend/"/><meta property="twitter:url" content="https://cobrexa.github.io/COBREXA.jl/stable/reference/frontend/"/><link rel="canonical" href="https://cobrexa.github.io/COBREXA.jl/stable/reference/frontend/"/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img class="docs-light-only" src="../../assets/logo.svg" alt="COBREXA.jl logo"/><img class="docs-dark-only" src="../../assets/logo-dark.svg" alt="COBREXA.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">COBREXA.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../../examples/">Contents</a></li><li><a class="tocitem" href="../../examples/01-loading-and-saving/">Loading and saving models</a></li><li><a class="tocitem" href="../../examples/02a-flux-balance-analysis/">Flux balance analysis (FBA)</a></li><li><a class="tocitem" href="../../examples/02b-optimizer-parameters/">Changing optimizer parameters</a></li><li><a class="tocitem" href="../../examples/02c-model-modifications/">Making adjustments to the model</a></li><li><a class="tocitem" href="../../examples/02d-constraint-modifications/">Making adjustments to the constraint system</a></li><li><a class="tocitem" href="../../examples/03a-flux-variability-analysis/">Flux variability analysis (FVA)</a></li><li><a class="tocitem" href="../../examples/03b-parsimonious-flux-balance/">Parsimonious flux balance analysis</a></li><li><a class="tocitem" href="../../examples/03c-envelopes/">Production envelopes</a></li><li><a class="tocitem" href="../../examples/03d-unidirectional/">Split unidirectional reactions in models</a></li><li><a class="tocitem" href="../../examples/04-community-models/">Community FBA models</a></li><li><a class="tocitem" href="../../examples/05a-minimization-of-metabolic-adjustment/">Minimization of metabolic adjustment analysis</a></li><li><a class="tocitem" href="../../examples/05b-enzyme-constrained-models/">Enzyme constrained models</a></li><li><a class="tocitem" href="../../examples/05c-mmdf/">Thermodynamic models</a></li><li><a class="tocitem" href="../../examples/05d-loopless-models/">Loopless flux balance analysis (ll-FBA)</a></li><li><a class="tocitem" href="../../examples/05e-knockouts/">Gene knockouts</a></li><li><a class="tocitem" href="../../examples/05f-cyclefree/">CycleFreeFlux</a></li><li><a class="tocitem" href="../../examples/05g-gapfilling/">Gap filling</a></li><li><a class="tocitem" href="../../examples/06a-sampling/">Flux sampling</a></li><li><a class="tocitem" href="../../examples/06b-screening/">Screening through many model variants</a></li></ul></li><li><a class="tocitem" href="../../structure/">Core concepts and structure</a></li><li><span class="tocitem">Parallel, distributed and HPC processing</span><ul><li><a class="tocitem" href="../../distributed/">Contents</a></li><li><a class="tocitem" href="../../distributed/1_functions/">Parallel processing overview</a></li><li><a class="tocitem" href="../../distributed/2_parallel/">Local parallel processing</a></li><li><a class="tocitem" href="../../distributed/3_slurm/">Working in a HPC environment</a></li></ul></li><li><span class="tocitem">API reference</span><ul><li><a class="tocitem" href="../">Contents</a></li><li><a class="tocitem" href="../analysis/">Specialized analysis functions</a></li><li><a class="tocitem" href="../builders/">Constraint system builders</a></li><li><a class="tocitem" href="../core/">Core functionality</a></li><li class="is-active"><a class="tocitem" href>Front-end user interface</a><ul class="internal"><li><a class="tocitem" href="#Flux-balance-analysis"><span>Flux balance analysis</span></a></li><li><a class="tocitem" href="#Flux-variability-analysis"><span>Flux variability analysis</span></a></li><li><a class="tocitem" href="#Parsimonious-flux-balance-analysis"><span>Parsimonious flux balance analysis</span></a></li><li><a class="tocitem" href="#Minimization-of-metabolic-adjustment"><span>Minimization of metabolic adjustment</span></a></li><li><a class="tocitem" href="#Enzyme-mass-constrained-models"><span>Enzyme-mass-constrained models</span></a></li><li><a class="tocitem" href="#Community-models"><span>Community models</span></a></li><li><a class="tocitem" href="#Production-envelopes"><span>Production envelopes</span></a></li><li><a class="tocitem" href="#Gap-filling"><span>Gap-filling</span></a></li><li><a class="tocitem" href="#Knockout-models"><span>Knockout models</span></a></li><li><a class="tocitem" href="#Loopless-flux-balance-analysis"><span>Loopless flux balance analysis</span></a></li><li><a class="tocitem" href="#Max-Min-Driving-Force-analysis"><span>Max-Min Driving Force analysis</span></a></li><li><a class="tocitem" href="#Sampling"><span>Sampling</span></a></li></ul></li><li><a class="tocitem" href="../misc/">Miscellaneous functions</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">API reference</a></li><li class="is-active"><a href>Front-end user interface</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Front-end user interface</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/COBREXA/COBREXA.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/COBREXA/COBREXA.jl/blob/master/docs/src/reference/frontend.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Front-end-user-interface"><a class="docs-heading-anchor" href="#Front-end-user-interface">Front-end user interface</a><a id="Front-end-user-interface-1"></a><a class="docs-heading-anchor-permalink" href="#Front-end-user-interface" title="Permalink"></a></h1><h2 id="Flux-balance-analysis"><a class="docs-heading-anchor" href="#Flux-balance-analysis">Flux balance analysis</a><a id="Flux-balance-analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Flux-balance-analysis" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}" href="#COBREXA.flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>COBREXA.flux_balance_analysis</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">flux_balance_analysis(
    model::AbstractFBCModels.AbstractFBCModel;
    kwargs...
)
</code></pre><p>Compute an optimal objective-optimizing solution of the given <code>model</code>.</p><p>Most arguments are forwarded to <a href="../analysis/#COBREXA.optimized_values-Tuple{ConstraintTrees.ConstraintTree}"><code>optimized_values</code></a>.</p><p>Returns a tree with the optimization solution of the same shape as given by <a href="#COBREXA.flux_balance_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>flux_balance_constraints</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/balance.jl#L131">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.flux_balance_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}" href="#COBREXA.flux_balance_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>COBREXA.flux_balance_constraints</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">flux_balance_constraints(
    model::AbstractFBCModels.AbstractFBCModel;
    interface,
    interface_name
) -&gt; Any
</code></pre><p>A constraint tree that models the content of the given instance of <code>AbstractFBCModel</code>.</p><p>The constructed tree contains subtrees <code>fluxes</code> (with the reaction-defining &quot;variables&quot;) and <code>flux_stoichiometry</code> (with the metabolite-balance-defining constraints), and a single constraint <code>objective</code> thad describes the objective function of the model.</p><p>Optionally if <code>interface</code> is specified, an &quot;interface block&quot; will be created within the constraint tree for later use as a &quot;module&quot; in creating bigger models (such as communities) using <a href="../builders/#COBREXA.interface_constraints-Tuple{Any}"><code>interface_constraints</code></a>. The possible parameter values include:</p><ul><li><code>nothing</code> – default, no interface is created</li><li><code>:sbo</code> – the interface gets created from model&#39;s SBO annotations)</li><li><code>:identifier_prefixes</code> – the interface is guesstimated from commonly occurring adhoc reaction ID prefixes used in contemporary models</li><li><code>:boundary</code> – the interface is created from all reactions that either only consume or only produce metabolites</li></ul><p>Output interface name can be set via <code>interface_name</code>.</p><p>See <a href="../core/#Configuration"><code>Configuration</code></a> for fine-tuning the default interface creation.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/balance.jl#L17">source</a></section></article><h2 id="Flux-variability-analysis"><a class="docs-heading-anchor" href="#Flux-variability-analysis">Flux variability analysis</a><a id="Flux-variability-analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Flux-variability-analysis" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.flux_variability_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}" href="#COBREXA.flux_variability_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>COBREXA.flux_variability_analysis</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">flux_variability_analysis(
    model::AbstractFBCModels.AbstractFBCModel;
    objective_bound,
    reactions,
    optimizer,
    settings,
    workers
)
</code></pre><p>Perform a Flux Variability Analysis (FVA) on the <code>model</code>, and return a dictionary of flux ranges where the model is able to perform optimally.</p><p>The constraint system is constructed using <a href="#COBREXA.flux_balance_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>flux_balance_constraints</code></a>, and the variability is examined on all reaction&#39;s fluxes, or on the subset given optionally in <code>reaction_subset</code> (e.g., <code>reaction_subset = [&quot;PFK&quot;, &quot;ACALD&quot;]</code>). The optimality tolerance can be specified with <code>objective_bound</code> using e.g. <a href="../misc/#COBREXA.relative_tolerance_bound-Tuple{Any}"><code>relative_tolerance_bound</code></a> or <a href="../misc/#COBREXA.absolute_tolerance_bound-Tuple{Any}"><code>absolute_tolerance_bound</code></a>; the default is 99% relative tolerance.</p><p>Parameter <code>workers</code> may be used to enable parallel or distributed processing; the execution defaults to all available workers. Other parameters (esp. <code>optimizer</code>) are internally forwarded to <a href="../analysis/#COBREXA.optimized_values-Tuple{ConstraintTrees.ConstraintTree}"><code>optimized_values</code></a>.</p><p>Use <a href="../analysis/#COBREXA.constraints_variability-Tuple{ConstraintTrees.ConstraintTree, ConstraintTrees.ConstraintTree}"><code>constraints_variability</code></a> to customize the FVA execution.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/variability.jl#L17">source</a></section></article><h2 id="Parsimonious-flux-balance-analysis"><a class="docs-heading-anchor" href="#Parsimonious-flux-balance-analysis">Parsimonious flux balance analysis</a><a id="Parsimonious-flux-balance-analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Parsimonious-flux-balance-analysis" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.linear_parsimonious_flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}" href="#COBREXA.linear_parsimonious_flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>COBREXA.linear_parsimonious_flux_balance_analysis</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">linear_parsimonious_flux_balance_analysis(
    model::AbstractFBCModels.AbstractFBCModel;
    tolerances,
    kwargs...
)
</code></pre><p>Like <a href="#COBREXA.parsimonious_flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>parsimonious_flux_balance_analysis</code></a>, but uses the L1-metric parsimonious system given by <a href="#COBREXA.linear_parsimonious_flux_balance_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>linear_parsimonious_flux_balance_constraints</code></a>.</p><p>In turn, the solution is often faster, does not require a solver capable of quadratic objectives, and has many beneficial properties of the usual parsimonious solutions (such as the general lack of unnecessary loops). On the other hand, like with plain flux balance analysis there is no strong guarantee of uniqueness of the solution.</p><p>Solver configuration arguments are forwarded to <a href="../analysis/#COBREXA.parsimonious_optimized_values-Tuple{Union{ConstraintTrees.Constraint, ConstraintTrees.ConstraintTree}}"><code>parsimonious_optimized_values</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/parsimonious.jl#L91">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.linear_parsimonious_flux_balance_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}" href="#COBREXA.linear_parsimonious_flux_balance_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>COBREXA.linear_parsimonious_flux_balance_constraints</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">linear_parsimonious_flux_balance_constraints(
    model::AbstractFBCModels.AbstractFBCModel;
    kwargs...
) -&gt; ConstraintTrees.Tree
</code></pre><p>Like <a href="#COBREXA.parsimonious_flux_balance_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>parsimonious_flux_balance_constraints</code></a>, but uses a L1 metric for solving the parsimonious problem. The <code>parsimonious_objective</code> constraint is thus linear.</p><p>Keyword arguments are forwarded to <a href="#COBREXA.flux_balance_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>flux_balance_constraints</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/parsimonious.jl#L65">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.parsimonious_flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}" href="#COBREXA.parsimonious_flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>COBREXA.parsimonious_flux_balance_analysis</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">parsimonious_flux_balance_analysis(
    model::AbstractFBCModels.AbstractFBCModel;
    tolerances,
    kwargs...
)
</code></pre><p>Compute a parsimonious flux solution for the <code>model</code>, using the constraints given by <a href="#COBREXA.parsimonious_flux_balance_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>parsimonious_flux_balance_constraints</code></a>.</p><p>In short, the objective value of the parsimonious solution should be the same as the one from <a href="#COBREXA.flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>flux_balance_analysis</code></a>, except the squared sum of reaction fluxes is minimized. If there are multiple possible fluxes that achieve a given objective value, parsimonious flux thus represents the &quot;minimum energy&quot; one, which is arguably more realistic.</p><p>Solver configuration arguments are forwarded to <a href="../analysis/#COBREXA.parsimonious_optimized_values-Tuple{Union{ConstraintTrees.Constraint, ConstraintTrees.ConstraintTree}}"><code>parsimonious_optimized_values</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/parsimonious.jl#L35">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.parsimonious_flux_balance_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}" href="#COBREXA.parsimonious_flux_balance_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>COBREXA.parsimonious_flux_balance_constraints</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">parsimonious_flux_balance_constraints(
    model::AbstractFBCModels.AbstractFBCModel;
    kwargs...
) -&gt; Any
</code></pre><p>A constraint system like from <a href="#COBREXA.flux_balance_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>flux_balance_constraints</code></a>, but with the parsimonious objective present under key <code>parsimonious_objective</code>. Best used via <a href="#COBREXA.parsimonious_flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>parsimonious_flux_balance_analysis</code></a>.</p><p>Keyword arguments are forwarded to <a href="#COBREXA.flux_balance_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>flux_balance_constraints</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/parsimonious.jl#L17">source</a></section></article><h2 id="Minimization-of-metabolic-adjustment"><a class="docs-heading-anchor" href="#Minimization-of-metabolic-adjustment">Minimization of metabolic adjustment</a><a id="Minimization-of-metabolic-adjustment-1"></a><a class="docs-heading-anchor-permalink" href="#Minimization-of-metabolic-adjustment" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.linear_metabolic_adjustment_minimization_analysis-Tuple{AbstractFBCModels.AbstractFBCModel, Vararg{Any}}" href="#COBREXA.linear_metabolic_adjustment_minimization_analysis-Tuple{AbstractFBCModels.AbstractFBCModel, Vararg{Any}}"><code>COBREXA.linear_metabolic_adjustment_minimization_analysis</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">linear_metabolic_adjustment_minimization_analysis(
    model::AbstractFBCModels.AbstractFBCModel,
    args...;
    optimizer,
    settings,
    reference_parsimonious_optimizer,
    reference_parsimonious_settings,
    reference_optimizer,
    reference_settings,
    kwargs...
)
</code></pre><p>Perform a linear minimization of metabolic adjustment analysis (l-MOMA) on <code>model</code>. The reference is given by the second argument, which is either a <code>reference_flux</code> or a <code>reference_model</code> (the second argument is forwarded to <a href="#COBREXA.linear_metabolic_adjustment_minimization_constraints-Tuple{AbstractFBCModels.AbstractFBCModel, AbstractFBCModels.AbstractFBCModel}"><code>linear_metabolic_adjustment_minimization_constraints</code></a>).</p><p>While the solution is &quot;less uniquely defined&quot; than with fully quadratic <a href="#COBREXA.metabolic_adjustment_minimization_analysis-Tuple{AbstractFBCModels.AbstractFBCModel, Vararg{Any}}"><code>metabolic_adjustment_minimization_analysis</code></a>, the linear variant typically produces a sufficiently good result with much less resources. See documentation of <a href="#COBREXA.linear_parsimonious_flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>linear_parsimonious_flux_balance_analysis</code></a> for some of the considerations.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/moma.jl#L183">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.linear_metabolic_adjustment_minimization_constraints-Tuple{AbstractFBCModels.AbstractFBCModel, AbstractFBCModels.AbstractFBCModel}" href="#COBREXA.linear_metabolic_adjustment_minimization_constraints-Tuple{AbstractFBCModels.AbstractFBCModel, AbstractFBCModels.AbstractFBCModel}"><code>COBREXA.linear_metabolic_adjustment_minimization_constraints</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">linear_metabolic_adjustment_minimization_constraints(
    model::AbstractFBCModels.AbstractFBCModel,
    reference_model::AbstractFBCModels.AbstractFBCModel;
    kwargs...
) -&gt; Union{Nothing, ConstraintTrees.Tree}
</code></pre><p>Like <a href="#COBREXA.metabolic_adjustment_minimization_constraints-Tuple{AbstractFBCModels.AbstractFBCModel, AbstractFBCModels.AbstractFBCModel}"><code>metabolic_adjustment_minimization_constraints</code></a> but the output constraints optimize the L1 distance from the linear-parsimonious solution of the <code>reference_model</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/moma.jl#L156">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.linear_metabolic_adjustment_minimization_constraints-Tuple{AbstractFBCModels.AbstractFBCModel, ConstraintTrees.Tree}" href="#COBREXA.linear_metabolic_adjustment_minimization_constraints-Tuple{AbstractFBCModels.AbstractFBCModel, ConstraintTrees.Tree}"><code>COBREXA.linear_metabolic_adjustment_minimization_constraints</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">linear_metabolic_adjustment_minimization_constraints(
    model::AbstractFBCModels.AbstractFBCModel,
    reference_fluxes::ConstraintTrees.Tree;
    _...
) -&gt; ConstraintTrees.Tree
</code></pre><p>Like <a href="#COBREXA.metabolic_adjustment_minimization_constraints-Tuple{AbstractFBCModels.AbstractFBCModel, AbstractFBCModels.AbstractFBCModel}"><code>metabolic_adjustment_minimization_constraints</code></a> but optimizes the L1 distance from <code>reference_fluxes</code>.</p><p>Keyword arguments are discarded for compatibility with the other overload.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/moma.jl#L118">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.metabolic_adjustment_minimization_analysis-Tuple{AbstractFBCModels.AbstractFBCModel, Vararg{Any}}" href="#COBREXA.metabolic_adjustment_minimization_analysis-Tuple{AbstractFBCModels.AbstractFBCModel, Vararg{Any}}"><code>COBREXA.metabolic_adjustment_minimization_analysis</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">metabolic_adjustment_minimization_analysis(
    model::AbstractFBCModels.AbstractFBCModel,
    args...;
    optimizer,
    settings,
    reference_parsimonious_optimizer,
    reference_parsimonious_settings,
    reference_optimizer,
    reference_settings,
    kwargs...
)
</code></pre><p>Find a solution of the &quot;minimization of metabolic adjustment&quot; (MOMA) analysis for the <code>model</code>, which is the &quot;closest&quot; feasible solution to the solution given in the second argument, which is either <code>reference_fluxes</code> or <code>reference_model</code> (see documentation of <a href="#COBREXA.metabolic_adjustment_minimization_constraints-Tuple{AbstractFBCModels.AbstractFBCModel, AbstractFBCModels.AbstractFBCModel}"><code>metabolic_adjustment_minimization_constraints</code></a>), in the sense of squared-sum distance. The minimized squared distance (the objective) is present in the result tree as <code>minimal_adjustment_objective</code>.</p><p>If the second argument is a reference model, it is solved using a <a href="#COBREXA.parsimonious_flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>parsimonious_flux_balance_analysis</code></a> with the optimizer and settings parameters for the 2 steps set by keyword arguments prefixed by <code>reference_</code>.</p><p>This is often used for models with smaller feasible region than the reference models (typically handicapped by a knockout, a nutritional deficiency or a similar perturbation). MOMA solution then gives an expectable &quot;easiest&quot; adjustment of the organism towards a somewhat working state.</p><p>Reference fluxes that do not exist in the <code>model</code> are ignored (internally, the objective is constructed via <a href="../builders/#COBREXA.squared_sum_error_value-Tuple{ConstraintTrees.ConstraintTree, Any}"><code>squared_sum_error_value</code></a>).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/moma.jl#L67">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.metabolic_adjustment_minimization_constraints-Tuple{AbstractFBCModels.AbstractFBCModel, AbstractFBCModels.AbstractFBCModel}" href="#COBREXA.metabolic_adjustment_minimization_constraints-Tuple{AbstractFBCModels.AbstractFBCModel, AbstractFBCModels.AbstractFBCModel}"><code>COBREXA.metabolic_adjustment_minimization_constraints</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">metabolic_adjustment_minimization_constraints(
    model::AbstractFBCModels.AbstractFBCModel,
    reference_model::AbstractFBCModels.AbstractFBCModel;
    kwargs...
) -&gt; Any
</code></pre><p>A slightly easier-to-use version of <a href="#COBREXA.metabolic_adjustment_minimization_constraints-Tuple{AbstractFBCModels.AbstractFBCModel, AbstractFBCModels.AbstractFBCModel}"><code>metabolic_adjustment_minimization_constraints</code></a> that computes the reference flux as the parsimonious optimal solution of the <code>reference_model</code>. The reference flux is calculated using <code>reference_optimizer</code> and <code>reference_modifications</code>, which default to the <code>optimizer</code> and <code>settings</code>.</p><p>Other arguments are forwarded to the internal call of <a href="../analysis/#COBREXA.parsimonious_optimized_values-Tuple{Union{ConstraintTrees.Constraint, ConstraintTrees.ConstraintTree}}"><code>parsimonious_optimized_values</code></a>.</p><p>Returns <code>nothing</code> if no feasible solution is found.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/moma.jl#L34">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.metabolic_adjustment_minimization_constraints-Tuple{AbstractFBCModels.AbstractFBCModel, ConstraintTrees.Tree}" href="#COBREXA.metabolic_adjustment_minimization_constraints-Tuple{AbstractFBCModels.AbstractFBCModel, ConstraintTrees.Tree}"><code>COBREXA.metabolic_adjustment_minimization_constraints</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">metabolic_adjustment_minimization_constraints(
    model::AbstractFBCModels.AbstractFBCModel,
    reference_fluxes::ConstraintTrees.Tree;
    _...
) -&gt; Any
</code></pre><p>Keyword arguments are discarded for compatibility with the other overload.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/moma.jl#L17">source</a></section></article><h3 id="Constraint-systems-for-metabolite-concentrations"><a class="docs-heading-anchor" href="#Constraint-systems-for-metabolite-concentrations">Constraint systems for metabolite concentrations</a><a id="Constraint-systems-for-metabolite-concentrations-1"></a><a class="docs-heading-anchor-permalink" href="#Constraint-systems-for-metabolite-concentrations" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.log_concentration_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}" href="#COBREXA.log_concentration_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>COBREXA.log_concentration_constraints</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">log_concentration_constraints(
    model::AbstractFBCModels.AbstractFBCModel;
    reactions,
    metabolites,
    metabolite_concentration_bound,
    reaction_concentration_bound
) -&gt; ConstraintTrees.ConstraintTree
</code></pre><p>Build log-concentration-stoichiometry constraints for the <code>model</code>, as used e.g. by <a href="#COBREXA.max_min_driving_force_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>max_min_driving_force_analysis</code></a>.</p><p>The output constraint tree contains a log-concentration variable for each <code>metabolite</code> from <code>model</code>, in subtree <code>log_concentrations</code>. The total reactant log-concentrations for each <code>reaction</code> are constrained in subtree <code>log_concentration_stoichiometry</code>. By default, all reactions and metabolites in <code>model</code> are included.</p><p>A concentration bound is given by parameter function <code>concentration_bound</code> for each metabolite ID (the string ID is the single argument of the function); by default the function returns <code>nothing</code> and no bounds are installed. The same is used for reactions with <code>reaction_concentration_bound</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/concentrations.jl#L17">source</a></section></article><h2 id="Enzyme-mass-constrained-models"><a class="docs-heading-anchor" href="#Enzyme-mass-constrained-models">Enzyme-mass-constrained models</a><a id="Enzyme-mass-constrained-models-1"></a><a class="docs-heading-anchor-permalink" href="#Enzyme-mass-constrained-models" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.Isozyme" href="#COBREXA.Isozyme"><code>COBREXA.Isozyme</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">mutable struct IsozymeT{Float64}</code></pre><p>Shortcut for <code>[IsozymeT](@ref){Float64}</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/enzymes.jl#L45">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.IsozymeT" href="#COBREXA.IsozymeT"><code>COBREXA.IsozymeT</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">mutable struct IsozymeT{T&lt;:Real}</code></pre><p>A simple struct storing information about the isozyme composition, including subunit stoichiometry and turnover numbers. Use with <a href="#COBREXA.enzyme_constrained_flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>enzyme_constrained_flux_balance_analysis</code></a>.</p><p><strong>Fields</strong></p><ul><li><code>gene_product_stoichiometry::Dict{String, T} where T&lt;:Real</code>: Mapping of gene product identifiers (&quot;genes&quot; in FBC model nomenclature) to their relative amount required to construct one unit of the isozyme.</li></ul><ul><li><p><code>kcat_forward::Union{Nothing, T} where T&lt;:Real</code>: Turnover number for this isozyme catalyzing the forward direction of the     reaction.</p></li><li><p><code>kcat_reverse::Union{Nothing, T} where T&lt;:Real</code>: Turnover number for this isozyme catalyzing the reverse direction of the     reaction.</p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/enzymes.jl#L17">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.enzyme_constrained_flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}" href="#COBREXA.enzyme_constrained_flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>COBREXA.enzyme_constrained_flux_balance_analysis</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">enzyme_constrained_flux_balance_analysis(
    model::AbstractFBCModels.AbstractFBCModel;
    kwargs...
)
</code></pre><p>Perform the enzyme-constrained flux balance analysis on the <code>model</code> and return the solved constraint system.</p><p>Arguments are forwarded to <a href="#COBREXA.enzyme_constrained_flux_balance_constraints-Union{Tuple{AbstractFBCModels.AbstractFBCModel}, Tuple{R}} where R&lt;:Real"><code>enzyme_constrained_flux_balance_constraints</code></a>; solver configuration arguments are forwarded to <a href="../analysis/#COBREXA.optimized_values-Tuple{ConstraintTrees.ConstraintTree}"><code>optimized_values</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/enzymes.jl#L221">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.enzyme_constrained_flux_balance_constraints-Union{Tuple{AbstractFBCModels.AbstractFBCModel}, Tuple{R}} where R&lt;:Real" href="#COBREXA.enzyme_constrained_flux_balance_constraints-Union{Tuple{AbstractFBCModels.AbstractFBCModel}, Tuple{R}} where R&lt;:Real"><code>COBREXA.enzyme_constrained_flux_balance_constraints</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">enzyme_constrained_flux_balance_constraints(
    model::AbstractFBCModels.AbstractFBCModel;
    reaction_isozymes,
    gene_product_molar_masses,
    capacity,
    interface,
    interface_name
)
</code></pre><p>Construct a enzyme-constrained flux-balance constraint system, following the method in GECKO algorithm (refer to: <em>Sánchez, Benjamín J., et al. &quot;Improving the phenotype predictions of a yeast genome‐scale metabolic model by incorporating enzymatic constraints.&quot; Molecular systems biology 13.8 (2017): 935</em>).</p><p>The enzyme mass constraints depend primarily on the available <em>isozymes</em>, given in parameter <code>reaction_isozymes</code>, which is a mapping of reaction identifiers to descriptions of <a href="#COBREXA.Isozyme"><code>Isozyme</code></a>s that may catalyze the particular reactions. The isozymes are built from gene products, the mass of which is specified by <code>gene_product_molar_masses</code>. In total, the amount of gene product building material is limited by <code>capacity</code>.</p><p><code>capacity</code> may be a single number, which sets the mass limit for &quot;all described enzymes&quot;. Alternatively, <code>capacity</code> may be a vector of identifier-genes-limit triples that together form a constraint (identified by the given identifier) that limits the total sum of the listed genes to the given limit. The interpretation of <code>capacity</code> is implemented (and can be extended) via <a href="#COBREXA.expand_enzyme_capacity-Tuple{Any, Any}"><code>expand_enzyme_capacity</code></a>.</p><p><code>interface</code> and <code>interface_name</code> are forwarded to <a href="#COBREXA.flux_balance_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>flux_balance_constraints</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/enzymes.jl#L127">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.expand_enzyme_capacity-Tuple{Any, Any}" href="#COBREXA.expand_enzyme_capacity-Tuple{Any, Any}"><code>COBREXA.expand_enzyme_capacity</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">expand_enzyme_capacity(x, all) -&gt; Vector
</code></pre><p>Expand the <code>capacity</code> argument as given to <a href="#COBREXA.enzyme_constrained_flux_balance_constraints-Union{Tuple{AbstractFBCModels.AbstractFBCModel}, Tuple{R}} where R&lt;:Real"><code>enzyme_constrained_flux_balance_constraints</code></a> and <a href="#COBREXA.simplified_enzyme_constrained_flux_balance_constraints-Union{Tuple{Any}, Tuple{R}} where R&lt;:Real"><code>simplified_enzyme_constrained_flux_balance_constraints</code></a> into a form accepted by <a href="../builders/#COBREXA.enzyme_constraints-Tuple{}"><code>enzyme_constraints</code></a> and <a href="../builders/#COBREXA.simplified_enzyme_constraints-Tuple{}"><code>simplified_enzyme_constraints</code></a> (respectively).</p><p>By default, <code>Bound</code>s are kept intact, <code>Real</code> values are converted to a fixed interval between a zero and the value. All other values are assumed to be lists of capacities. (See <a href="#COBREXA.expand_enzyme_capacity_bound-Tuple{ConstraintTrees.Bound}"><code>expand_enzyme_capacity_bound</code></a> for translation of actual bounds).</p><p>Overloading this function (or <a href="#COBREXA.expand_enzyme_capacity_bound-Tuple{ConstraintTrees.Bound}"><code>expand_enzyme_capacity_bound</code></a>) gives a way to simplify the interface of the functions by accomodating custom capacity types.</p><p>The second argument is provided to this function as a list of full scope of the capacities it can work with, by default &quot;all capacities&quot;.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/enzymes.jl#L55">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.expand_enzyme_capacity-Tuple{Union{Dict, Vector{Pair}}, Any}" href="#COBREXA.expand_enzyme_capacity-Tuple{Union{Dict, Vector{Pair}}, Any}"><code>COBREXA.expand_enzyme_capacity</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">expand_enzyme_capacity(
    x::Union{Dict, Vector{Pair}},
    _
) -&gt; Vector
</code></pre><p>Overload of <a href="#COBREXA.expand_enzyme_capacity-Tuple{Any, Any}"><code>expand_enzyme_capacity</code></a> for all <code>Dict</code>-like iterables.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/enzymes.jl#L79">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.expand_enzyme_capacity-Tuple{Vector{&lt;:Tuple}, Any}" href="#COBREXA.expand_enzyme_capacity-Tuple{Vector{&lt;:Tuple}, Any}"><code>COBREXA.expand_enzyme_capacity</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">expand_enzyme_capacity(x::Vector{&lt;:Tuple}, _) -&gt; Vector
</code></pre><p>Overload of <a href="#COBREXA.expand_enzyme_capacity-Tuple{Any, Any}"><code>expand_enzyme_capacity</code></a> that provides compatibility with the earlier capacity specifications (using triples instead of pairs).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/enzymes.jl#L87">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.expand_enzyme_capacity_bound-Tuple{ConstraintTrees.Bound}" href="#COBREXA.expand_enzyme_capacity_bound-Tuple{ConstraintTrees.Bound}"><code>COBREXA.expand_enzyme_capacity_bound</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">expand_enzyme_capacity_bound(
    x::ConstraintTrees.Bound
) -&gt; ConstraintTrees.Bound
</code></pre><p>By default, <a href="#COBREXA.expand_enzyme_capacity_bound-Tuple{ConstraintTrees.Bound}"><code>expand_enzyme_capacity_bound</code></a> leaves all <code>Bound</code>s intact. This overload implements this property.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/enzymes.jl#L117">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.expand_enzyme_capacity_bound-Tuple{Real}" href="#COBREXA.expand_enzyme_capacity_bound-Tuple{Real}"><code>COBREXA.expand_enzyme_capacity_bound</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">expand_enzyme_capacity_bound(x::Real) -&gt; Tuple{Real, Real}
</code></pre><p>Expand a single capacity bound for use in enzyme-constrained models. Overloading this function provides additional ways to interpret the capacity specifications. Typically used via <a href="#COBREXA.expand_enzyme_capacity-Tuple{Any, Any}"><code>expand_enzyme_capacity</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/enzymes.jl#L108">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.expand_enzyme_capacity_iterable-Tuple{Any}" href="#COBREXA.expand_enzyme_capacity_iterable-Tuple{Any}"><code>COBREXA.expand_enzyme_capacity_iterable</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">expand_enzyme_capacity_iterable(x) -&gt; Any
</code></pre><p>Internal helper for implementation of <a href="#COBREXA.expand_enzyme_capacity-Tuple{Any, Any}"><code>expand_enzyme_capacity</code></a> over iterable <code>Dict</code>-like objects.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/enzymes.jl#L98">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.simplified_enzyme_constrained_flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}" href="#COBREXA.simplified_enzyme_constrained_flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>COBREXA.simplified_enzyme_constrained_flux_balance_analysis</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">simplified_enzyme_constrained_flux_balance_analysis(
    model::AbstractFBCModels.AbstractFBCModel;
    kwargs...
)
</code></pre><p>Perform the enzyme-constrained flux balance analysis on the <code>model</code> and return the solved constraint system.</p><p>Arguments are forwarded to <a href="#COBREXA.simplified_enzyme_constrained_flux_balance_constraints-Union{Tuple{Any}, Tuple{R}} where R&lt;:Real"><code>simplified_enzyme_constrained_flux_balance_constraints</code></a>; solver configuration arguments are forwarded to <a href="../analysis/#COBREXA.optimized_values-Tuple{ConstraintTrees.ConstraintTree}"><code>optimized_values</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/enzymes.jl#L361">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.simplified_enzyme_constrained_flux_balance_constraints-Union{Tuple{Any}, Tuple{R}} where R&lt;:Real" href="#COBREXA.simplified_enzyme_constrained_flux_balance_constraints-Union{Tuple{Any}, Tuple{R}} where R&lt;:Real"><code>COBREXA.simplified_enzyme_constrained_flux_balance_constraints</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">simplified_enzyme_constrained_flux_balance_constraints(
    model;
    reaction_isozymes,
    gene_product_molar_masses,
    capacity,
    interface,
    interface_name
)
</code></pre><p>Like <a href="#COBREXA.enzyme_constrained_flux_balance_constraints-Union{Tuple{AbstractFBCModels.AbstractFBCModel}, Tuple{R}} where R&lt;:Real"><code>enzyme_constrained_flux_balance_constraints</code></a>, but automatically selects a single &quot;fastest&quot; isozyme for each reaction direction. In turn, the system requires much less variables in the constraint system description, and usually solves more efficiently, for the price of possibly finding suboptimal solutions. The method follows the SMOMENT algorithm described in <em>Bekiaris, P.S., Klamt, S. Automatic construction of metabolic models with enzyme constraints. BMC Bioinformatics 21, 19 (2020). https://doi.org/10.1186/s12859-019-3329-9</em>.</p><p>Arguments are as with <a href="#COBREXA.enzyme_constrained_flux_balance_constraints-Union{Tuple{AbstractFBCModels.AbstractFBCModel}, Tuple{R}} where R&lt;:Real"><code>enzyme_constrained_flux_balance_constraints</code></a>, with a major difference in <code>capacity</code> handling: the identifier lists contain reactions identifiers (i.e., keys of <code>fluxes</code> in the constraint tree), instead of the gene product identifiers.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/enzymes.jl#L241">source</a></section></article><h2 id="Community-models"><a class="docs-heading-anchor" href="#Community-models">Community models</a><a id="Community-models-1"></a><a class="docs-heading-anchor-permalink" href="#Community-models" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.community_flux_balance_analysis-Tuple" href="#COBREXA.community_flux_balance_analysis-Tuple"><code>COBREXA.community_flux_balance_analysis</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">community_flux_balance_analysis(args...; kwargs...)
</code></pre><p>Run the Community Flux Balance Analysis on several models. All arguments are forwarded to <a href="#COBREXA.community_flux_balance_constraints-Tuple{Any, Any}"><code>community_flux_balance_constraints</code></a> which constructs the model; this function returns the solved model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/community.jl#L88">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.community_flux_balance_constraints-Tuple{Any, Any}" href="#COBREXA.community_flux_balance_constraints-Tuple{Any, Any}"><code>COBREXA.community_flux_balance_constraints</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">community_flux_balance_constraints(
    model_abundances,
    community_exchange_bounds;
    interface,
    interface_exchanges,
    interface_biomass,
    default_community_exchange_bound
) -&gt; ConstraintTrees.ConstraintTree
</code></pre><p>Construct an instance of a linear community Flux Balance Analysis (cFBA) model. The relative abundances of the organisms are known in advance; this function predicts the maximum achievable community growth rate.</p><p><code>model_abundances</code> is a dictionary-like object that maps model identifiers to tuples of models (usually subtypes of <code>AbstractFBCModel</code>) and their abundances (such as: <code>&quot;bug1&quot; =&gt; (bug1, 0.5), ...</code>). <code>community_exchange_bounds</code> is a dictionary-like object that can be additionally used to restrict selected community exchange reactions (keys should be reaction IDs, the values are converted to <code>ConstraintTrees</code>-like bounds). Bounds otherwise default to parameter <code>default_community_exchange_bound</code>, which itself defaults to <code>nothing</code> (i.e., unbounded).</p><p>If required, constraint trees may be supplied instead of <code>AbstracFBCModel</code>s in <code>model_abundances</code>. These must provide an interface compatible with <code>interface_exchanges</code> and <code>interface_biomass</code>.</p><p><code>interface</code> is forwarded to <a href="#COBREXA.flux_balance_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>flux_balance_constraints</code></a>. <code>interface_exchanges</code> and <code>interface_biomass</code> are used to pick up the correct interface part to contribute to the community exchanges and community biomass.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/community.jl#L21">source</a></section></article><h2 id="Production-envelopes"><a class="docs-heading-anchor" href="#Production-envelopes">Production envelopes</a><a id="Production-envelopes-1"></a><a class="docs-heading-anchor-permalink" href="#Production-envelopes" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.objective_production_envelope-Tuple{AbstractFBCModels.AbstractFBCModel, Vector{String}}" href="#COBREXA.objective_production_envelope-Tuple{AbstractFBCModels.AbstractFBCModel, Vector{String}}"><code>COBREXA.objective_production_envelope</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">objective_production_envelope(
    model::AbstractFBCModels.AbstractFBCModel,
    reactions::Vector{String};
    breaks,
    optimizer,
    settings,
    workers
)
</code></pre><p>Find the objective production envelope of the <code>model</code> in the dimensions given by <code>reactions</code>.</p><p>This runs a variability analysis of the fluxes given by <a href="#COBREXA.flux_balance_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>flux_balance_constraints</code></a> to determine an applicable range for the dimensions, then splits the dimensions to equal-sized breaks (of count <code>breaks</code> for each dimension, i.e. total <code>breaks ^ length(reactions)</code> individual &quot;multidimensional breaks&quot;) thus forming a grid, and returns an array of fluxes through the model objective with the individual reactions fixed to flux as given by the grid.</p><p><code>optimizer</code> and <code>settings</code> are used to construct the optimization models.</p><p>The computation is distributed to the specified <code>workers</code>, defaulting to all available workers.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/envelope.jl#L17">source</a></section></article><h2 id="Gap-filling"><a class="docs-heading-anchor" href="#Gap-filling">Gap-filling</a><a id="Gap-filling-1"></a><a class="docs-heading-anchor-permalink" href="#Gap-filling" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.gap_filling_analysis-Tuple" href="#COBREXA.gap_filling_analysis-Tuple"><code>COBREXA.gap_filling_analysis</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">gap_filling_analysis(args...; kwargs...)
</code></pre><p>Run the gap-filling analysis on a constraint system specified by <a href="#COBREXA.gap_filling_constraints-Tuple{AbstractFBCModels.AbstractFBCModel, AbstractFBCModels.AbstractFBCModel, Float64}"><code>gap_filling_constraints</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/gapfill.jl#L176">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.gap_filling_constraints-Tuple{AbstractFBCModels.AbstractFBCModel, AbstractFBCModels.AbstractFBCModel, Float64}" href="#COBREXA.gap_filling_constraints-Tuple{AbstractFBCModels.AbstractFBCModel, AbstractFBCModels.AbstractFBCModel, Float64}"><code>COBREXA.gap_filling_constraints</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">gap_filling_constraints(
    model::AbstractFBCModels.AbstractFBCModel,
    universal_model::AbstractFBCModels.AbstractFBCModel,
    objective_target::Float64;
    universal_reaction_cost,
    max_cost,
    known_fills,
    kwargs...
) -&gt; ConstraintTrees.ConstraintTree
</code></pre><p>Make a gap-filling system from a FBC model with gaps and an universal FBC model that contains reactions to be added into the original model.</p><p>The output system will be constrainted to reach at least <code>objective_target</code> flux through the objective function. Generally, this should be set to an arbitrary small value such as <code>0.05</code>.</p><p><code>universal_reaction_cost</code> should assign a numeric cost of inclusion of each of the reactions in the <code>universal_model</code>; by default all are assigned equal weight of <code>1</code>. <code>max_cost</code> puts an optional maximum limit on the cost, which may help the solver to avoid exploring unnecessarily complex solutions. <code>known_fills</code> may contain previous solutions of the same system; these will be made infeasible in the output constraint system in order to allow discovery of other ones.</p><p>Additional arguments are forwarded to <code>flux_balance_constraints</code> that converts <code>model</code> to constraints.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/gapfill.jl#L17">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.gap_filling_constraints-Tuple{}" href="#COBREXA.gap_filling_constraints-Tuple{}"><code>COBREXA.gap_filling_constraints</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">gap_filling_constraints(
;
    system,
    stoichiometry,
    universal_fluxes,
    universal_stoichiometry,
    flux_cost,
    max_cost,
    known_fills
)
</code></pre><p>Make a gap-fillign system from a non-solving <code>system</code> with a separated <code>stoichiometry</code>, filling in possible fluxes from <code>universal_fluxes</code> that are balanced with <code>universal_stoichiometry</code></p><p><code>flux_cost</code> can be used to assign a weight to each given universal flux; the total cost is bounded by <code>max_cost</code>.</p><p><code>known_fills</code> may contain previous solutions to be avoided; these are processed by <a href="#COBREXA.gap_filling_known_fill_constraint-Tuple{ConstraintTrees.ConstraintTree, ConstraintTrees.Tree{Float64}}"><code>gap_filling_known_fill_constraint</code></a> and attached to the system.</p><p><code>stoichiometry</code> needs to be extended to construct the final constraints, thus it should not be present in the original <code>system</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/gapfill.jl#L63">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.gap_filling_known_fill_constraint-Tuple{ConstraintTrees.ConstraintTree, ConstraintTrees.Tree{Float64}}" href="#COBREXA.gap_filling_known_fill_constraint-Tuple{ConstraintTrees.ConstraintTree, ConstraintTrees.Tree{Float64}}"><code>COBREXA.gap_filling_known_fill_constraint</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">gap_filling_known_fill_constraint(
    fill_flags::ConstraintTrees.ConstraintTree,
    known_flags::ConstraintTrees.Tree{Float64}
) -&gt; ConstraintTrees.Constraint
</code></pre><p>Produce a constraint that can be added to the system made by <a href="#COBREXA.gap_filling_constraints-Tuple{AbstractFBCModels.AbstractFBCModel, AbstractFBCModels.AbstractFBCModel, Float64}"><code>gap_filling_constraints</code></a> to avoid repeating of a solution that includes reactions already generated by another solution, as represented by the solved <code>fill_flags</code>.</p><p>Parameter <code>fill_flags</code> are the gapfilling flags of the given constraint system, parameter <code>known_flags</code> is expected to contain the solved <code>fill_flags</code> for the solution that is to be avoided.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/gapfill.jl#L149">source</a></section></article><h2 id="Knockout-models"><a class="docs-heading-anchor" href="#Knockout-models">Knockout models</a><a id="Knockout-models-1"></a><a class="docs-heading-anchor-permalink" href="#Knockout-models" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.gene_knockout_constraints-Tuple{ConstraintTrees.ConstraintTree, Any, AbstractFBCModels.AbstractFBCModel}" href="#COBREXA.gene_knockout_constraints-Tuple{ConstraintTrees.ConstraintTree, Any, AbstractFBCModels.AbstractFBCModel}"><code>COBREXA.gene_knockout_constraints</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">gene_knockout_constraints(
    fluxes::ConstraintTrees.ConstraintTree,
    knockout_genes,
    model::AbstractFBCModels.AbstractFBCModel
) -&gt; ConstraintTrees.ConstraintTree
</code></pre><p>Make a <code>ConstraintTree</code> that simulates a gene knockout of <code>knockout_genes</code> in the <code>model</code> and disables corresponding <code>fluxes</code> accordingly.</p><p>Keys of the fluxes must correspond to the reaction identifiers in the <code>model</code>.</p><p><code>knockout_genes</code> may be any collection that support element tests using <code>in</code>. Since the test is done many times, a <code>Set</code> is a preferred contained for longer lists of genes.</p><p>All constraints are equality constraints returned in a single flat <code>ConstraintTree</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/knockout.jl#L18">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.gene_knockout_constraints-Tuple{ConstraintTrees.ConstraintTree, String, AbstractFBCModels.AbstractFBCModel}" href="#COBREXA.gene_knockout_constraints-Tuple{ConstraintTrees.ConstraintTree, String, AbstractFBCModels.AbstractFBCModel}"><code>COBREXA.gene_knockout_constraints</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">gene_knockout_constraints(
    fluxes::ConstraintTrees.ConstraintTree,
    knockout_gene::String,
    model::AbstractFBCModels.AbstractFBCModel
) -&gt; ConstraintTrees.ConstraintTree
</code></pre><p>Convenience overload of <a href="#COBREXA.gene_knockout_constraints-Tuple{ConstraintTrees.ConstraintTree, Any, AbstractFBCModels.AbstractFBCModel}"><code>gene_knockout_constraints</code></a> for knocking out a single gene (without the necessity to store the gene identifier in a singleton container).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/knockout.jl#L46">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.gene_knockouts" href="#COBREXA.gene_knockouts"><code>COBREXA.gene_knockouts</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">gene_knockouts(
    model::AbstractFBCModels.AbstractFBCModel;
    ...
)
gene_knockouts(
    model::AbstractFBCModels.AbstractFBCModel,
    gene_combinations::Vector{&lt;:Union{String, NTuple{N, String} where N}};
    kwargs...
)
</code></pre><p>Compute the objective value of the <code>model</code> for all knockouts specified by <code>gene_combinations</code>, which is a vector of gene IDs or tuples of gene IDs that are knocked out in groups.</p><p>Returns a vector in the same order as <code>gene_combinations</code>.</p><p>Extra arguments (mainly, the <code>optimizer</code>) are forwarded to <a href="../analysis/#COBREXA.screen_optimization_model-Tuple{Any, ConstraintTrees.ConstraintTree, Vararg{Any}}"><code>screen_optimization_model</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/knockout.jl#L61">source</a></section></article><h2 id="Loopless-flux-balance-analysis"><a class="docs-heading-anchor" href="#Loopless-flux-balance-analysis">Loopless flux balance analysis</a><a id="Loopless-flux-balance-analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Loopless-flux-balance-analysis" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.loopless_flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}" href="#COBREXA.loopless_flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>COBREXA.loopless_flux_balance_analysis</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">loopless_flux_balance_analysis(
    model::AbstractFBCModels.AbstractFBCModel;
    kwargs...
)
</code></pre><p>Perform the loopless flux balance analysis on the <code>model</code>, returning the solved constraint system.</p><p>Arguments are forwarded to <a href="#COBREXA.loopless_flux_balance_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>loopless_flux_balance_constraints</code></a> (see the documentation for the description of the constraint system); solver configuration arguments are forwarded to <a href="../analysis/#COBREXA.optimized_values-Tuple{ConstraintTrees.ConstraintTree}"><code>optimized_values</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/loopless.jl#L74">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.loopless_flux_balance_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}" href="#COBREXA.loopless_flux_balance_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>COBREXA.loopless_flux_balance_constraints</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">loopless_flux_balance_constraints(
    model::AbstractFBCModels.AbstractFBCModel;
    flux_infinity_bound,
    driving_force_nonzero_bound,
    driving_force_infinity_bound
) -&gt; ConstraintTrees.ConstraintTree
</code></pre><p>Construct a flux-balance constraint system from <code>model</code> with added quasi-thermodynamic constraints that ensure that thermodynamically infeasible internal cycles can not occur. The method is closer described by: <em>Schellenberger, Lewis, and, Palsson.  &quot;Elimination of thermodynamically infeasible loops in steady-state metabolic models.&quot;, Biophysical journal, 2011`</em>.</p><p>The loopless condition comes with a performance overhead: the computation needs to find the null space of the stoichiometry matrix (essentially inverting it); and the subsequently created optimization problem contains binary variables for each internal reaction, thus requiring a MILP solver and a potentially exponential solving time.</p><p>Internally, the system is constructed by combining <a href="#COBREXA.flux_balance_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>flux_balance_constraints</code></a> and <a href="../builders/#COBREXA.loopless_constraints-Tuple{}"><code>loopless_constraints</code></a>.</p><p>The arguments <code>driving_force_max_bound</code> and <code>driving_force_nonzero_bound</code> set the bounds (possibly negated ones) on the virtual &quot;driving forces&quot; (G_i in the paper).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/loopless.jl#L17">source</a></section></article><h2 id="Max-Min-Driving-Force-analysis"><a class="docs-heading-anchor" href="#Max-Min-Driving-Force-analysis">Max-Min Driving Force analysis</a><a id="Max-Min-Driving-Force-analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Max-Min-Driving-Force-analysis" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.max_min_driving_force_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}" href="#COBREXA.max_min_driving_force_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>COBREXA.max_min_driving_force_analysis</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">max_min_driving_force_analysis(
    model::AbstractFBCModels.AbstractFBCModel;
    kwargs...
)
</code></pre><p>Perform the max-min driving force analysis on the <code>model</code>, returning the solved constraint system.</p><p>Arguments are forwarded to <a href="#COBREXA.max_min_driving_force_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>max_min_driving_force_constraints</code></a> (see the documentation for the description of the constraint system); solver configuration arguments are forwarded to <a href="../analysis/#COBREXA.optimized_values-Tuple{ConstraintTrees.ConstraintTree}"><code>optimized_values</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/mmdf.jl#L194">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.max_min_driving_force_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}" href="#COBREXA.max_min_driving_force_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>COBREXA.max_min_driving_force_constraints</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">max_min_driving_force_constraints(
    model::AbstractFBCModels.AbstractFBCModel;
    reaction_standard_gibbs_free_energies,
    reference_flux,
    concentration_ratios,
    constant_concentrations,
    ignored_metabolites,
    proton_metabolites,
    water_metabolites,
    concentration_lower_bound,
    concentration_upper_bound,
    T,
    R,
    reference_flux_atol
)
</code></pre><p>Create max-min driving force constraint system from <code>model</code>, using the supplied reaction standard Gibbs free energies in <code>reaction_standard_gibbs_free_energies</code>.</p><p>The method is described by: <em>Noor, et al., &quot;Pathway thermodynamics highlights kinetic obstacles in central metabolism.&quot;, PLoS computational biology, 2014.</em></p><p><code>reference_flux</code> sets the directions of each reaction in <code>model</code>. The scale of the values is not important, only the direction is examined (w.r.t. <code>reference_flux_atol</code> tolerance). Ideally, the supplied <code>reference_flux</code> should be completely free of internal cycles, which enables the thermodynamic consistency. To get the cycle-free flux, you can use <a href="#COBREXA.loopless_flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>loopless_flux_balance_analysis</code></a> (computationally demanding, but gives thermodynamically consistent solutions), <a href="#COBREXA.parsimonious_flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>parsimonious_flux_balance_analysis</code></a> or <a href="#COBREXA.linear_parsimonious_flux_balance_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>linear_parsimonious_flux_balance_analysis</code></a> (which is computationally simple, but the consistency is not guaranteed).</p><p>Internally, <a href="#COBREXA.log_concentration_constraints-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>log_concentration_constraints</code></a> is used to lay out the base structure of the problem.</p><p>Following arguments are set optionally:</p><ul><li><code>water_metabolites</code>, <code>proton_metabolites</code> and <code>ignored_metabolites</code> allow to completely ignore constraints on a part of the metabolite set, which is explicitly recommended especially for water and protons (since the analyses generally assume aqueous environment of constant pH)</li><li><code>constant_concentrations</code> can be used to fix the concentrations of the metabolites</li><li><code>concentration_lower_bound</code> and <code>concentration_upper_bound</code> set the default concentration bounds for all other metabolites</li><li><code>concentration ratios</code> is a dictionary that assigns a tuple of metabolite-metabolite-concentration ratio constraint names; e.g. ATP/ADP ratio can be fixed to five-times-more-ATP by setting <code>concentration_ratios = Dict(&quot;adenosin_ratio&quot; =&gt; (&quot;atp&quot;, &quot;adp&quot;, 5.0))</code></li><li><code>T</code> and <code>R</code> default to the usual required thermodynamic constraints in the expected units (the defaults assume the &quot;usual&quot; units, valuing 298.15 K and ~0.008314 kJ/K/mol, respectively). These multiply the log-concentrations to obtain the actual Gibbs energies, and thus driving forces.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/mmdf.jl#L17">source</a></section></article><h2 id="Sampling"><a class="docs-heading-anchor" href="#Sampling">Sampling</a><a id="Sampling-1"></a><a class="docs-heading-anchor-permalink" href="#Sampling" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="COBREXA.flux_sample-Tuple{AbstractFBCModels.AbstractFBCModel}" href="#COBREXA.flux_sample-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>COBREXA.flux_sample</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">flux_sample(
    model::AbstractFBCModels.AbstractFBCModel;
    seed,
    objective_bound,
    method,
    n_chains,
    collect_iterations,
    optimizer,
    settings,
    workers,
    kwargs...
)
</code></pre><p>Run a sampling algorithm on the near-optimal feasible space of the <code>model</code> (as specified by <code>objective_bound</code>). By default, the sampling algorithm is ACHR (the <code>method</code> parameter is defaulted to <a href="../analysis/#COBREXA.sample_chain_achr-Union{Tuple{SM}, Tuple{M}, Tuple{V}, Tuple{F}} where {F&lt;:Real, V&lt;:AbstractVector{F}, M&lt;:AbstractMatrix{F}, SM&lt;:AbstractMatrix{F}}"><code>sample_chain_achr</code></a>). The sampling algorithm is ran for <code>n_chains</code> and the iterations for collecting the sampled values are specified by <code>collect_iterations</code>.</p><p><code>optimizer</code> is used to generate the warmup (with <code>settings</code>) for the sampler using the usual unidimensional maximum-variability fluxes (as from <a href="#COBREXA.flux_variability_analysis-Tuple{AbstractFBCModels.AbstractFBCModel}"><code>flux_variability_analysis</code></a>). All computations are parallelized across <code>workers</code>.</p><p>Extra arguments are forwarded to <a href="../analysis/#COBREXA.sample_constraints-Union{Tuple{T}, Tuple{Function, ConstraintTrees.ConstraintTree}} where T"><code>sample_constraints</code></a>. Eventually the arguments will reach the <code>method</code> function, so extra arguments can be also used to customize the methods (e.g., by setting the <code>epsilon</code> for the ACHR sampler).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/COBREXA/COBREXA.jl/blob/ebdb562cf865b3835ebf8f9d98e0e8c7cd1c7f15/src/frontend/sample.jl#L17">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../core/">« Core functionality</a><a class="docs-footer-nextpage" href="../misc/">Miscellaneous functions »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.0 on <span class="colophon-date" title="Thursday 16 January 2025 09:35">Thursday 16 January 2025</span>. Using Julia version 1.11.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
